### DISCUSSION TITLE
Netmiko + TTP - Template File

### DISCUSSION THREAD

**Discussion by justrob1980** [ktbyers/netmiko#3513] [Q&A] [answered]
I was looking at the example of using Netmiko with TTP here > https://github.com/ktbyers/netmiko/blob/develop/examples/send_command_ttp.py

My question is, the example script has you putting in the TTP template raw into the script and then the script writes the template to an individual file. Is there a reason to do this and not just provide the template already in a separate file? Does the write process format the template in a special way?

**Comment by ktbyers** [comment, marked as answer]
I would generally just put the template in a separate file.

**Reply by justrob1980** [reply]
Awesome, I already have TTP templates I've tested against saved config files, next step is to go directly to the device and filter the show command with TTP, just wanted to make sure the templates would work. Thanks!

**Reply by justrob1980** [reply]
I am having an issue where the TTP template tests fine here -> https://textfsm.nornir.tech/
But when I pass it to the live device via Netmiko I get a result missing values. 

These are the test results
![textfsm nornir tech_results](https://github.com/user-attachments/assets/9967870a-5b89-45bb-98ae-093b50fd7474)

Here is the template on the box where the script is running 
<img width="529" alt="ShareX_kyWP4pY2QV" src="https://github.com/user-attachments/assets/86b9a4bf-aa13-4f5e-9689-f9eb745df762">

Here is the script:

from netmiko import ConnectHandler
from pprint import pprint

device = {
    "device_type": "cisco_xe",
    "host": "ip address",
    "username": "username",
    "password": "password",
}


command = "show redundancy application group 2"
with ConnectHandler(**device) as net_connect:
    output = net_connect.send_command(
        command, use_ttp=True, ttp_template="redun_template.ttp"
    )

print()
pprint(output)
print()

Here is what the script outputs
<img width="512" alt="netmiko_output" src="https://github.com/user-attachments/assets/16f8ba1e-f4aa-4a0b-aca2-b088e21eb7a8">

Any clues as to why the results are different? Did I miss something in the template syntax? If so, why does the template syntax work in the test site but not via Netmiko?

**Reply by ktbyers** [reply]
I would check:
1. Exact same versions of TTP
2. Raw output that you are using on the TTP site is identical to raw output from Netmiko.
3. You might need check tabs/spacing (i.e. copy-and-paste of content can possibly change this in web forms).

**Reply by justrob1980** [reply]
How would you suggest I see what Netmiko is specifically see, an instance of ConnectHandler without the ttp flag set output somehow?

**Reply by ktbyers** [reply]
Capture the output without sending it through TTP (i.e. just straight 'send_command' call) and then just print it out.

```python
output = net_connect.send_command(command)
```

**Reply by justrob1980** [reply]
This is the result

('Group ID:2\n'
'Group Name:ATL-VZB-CUBE-HA\n'
'\n'
'Administrative State: No Shutdown\n'
'Aggregate operational state : Up\n'
'My Role: ACTIVE\n'
'Peer Role: STANDBY\n'
'Peer Presence: Yes\n'
'Peer Comm: Yes\n'
'Peer Progression Started: Yes\n'
'\n'
'RF Domain: btob-two\n'
'\t RF state: ACTIVE\n'
'\t Peer RF state: STANDBY HOT\n')

I presume \t here for TAB, however when I delete all the spaces in the template and then insert a tab and run it, I still do not get RF state or Peer RF state when running through Netmiko.
